### PR TITLE
164829199 implement jwt authentication

### PR DESCRIPTION
**Pull request by ValerieRono**
**Description**
Currently, our application uses session based authentication. 
The new token-based authentication systems allow users to enter their email and password in order to obtain a token which allows them to fetch a specific resource - without entering their username and password at each request. Once their token has been obtained, the user can use the token to access specific resources for a set time period.
JWT is a self-contained token which has authentication information, expire time information, and other user defined claims digitally signed.

**Type of Change**
-New feature(non-breaking change which adds functionality)
-This change requires a documentation update 

**Checklist**
-My code follows the style guidelines of this project
-I have linted my code prior to submission
-I had added tests
-New and existing unit tests pass locally with my changes

Pivotal tracker
164829199

### CODE REVIEW

**Review by bl4ck4ndbr0wn — APPROVED** [review]

**Review by vino-jasuba — APPROVED** [review]

**Review by ogiste — APPROVED** [review]
LGTM.

File: authors/apps/authentication/tests/test_jwt.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+                'password': 'password'
+                }
+            }
+        self.user = User.objects.create(
+            username='user',
+            email='user@mail.com',
+            password='password'
+        )
+        self.login_data = {
+            'user': {
```
Comment (oagutu) [inline authors/apps/authentication/tests/test_jwt.py]: Considering `self.login_data` is similar to `self.user_data`, You don't need to have a separate dict, you can just use `user_data` when logging in

File: authors/apps/authentication/views.py [hunk 2]
```diff
@@ -25,8 +25,8 @@ def post(self, request):
         serializer = self.serializer_class(data=user)
         serializer.is_valid(raise_exception=True)
         serializer.save()
-
-        return Response(serializer.data, status=status.HTTP_201_CREATED)
+        token = serializer.instance.token
+        return Response({'token': token}, status=status.HTTP_201_CREATED)
```
Comment (oagutu) [inline authors/apps/authentication/views.py]: Please include a success message along with the token. Also do this for `line 48`

File: authors/apps/authentication/tests/test_jwt.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+            reverse('authentication:register'),
+            self.user_data,
+            format='json'
+            )
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+        self.assertIn('token', response.data)
+
+    def test_user_gets_a_token_when_they_log_in(self):
+        """ Users should get a token when they successfully log in """
+        self.client.post(
```
Comment (oagutu) [inline authors/apps/authentication/tests/test_jwt.py]: Why not just use the already registered user and skip the registration step here?

File: authors/apps/authentication/backends.py [hunk 4]
```diff
@@ -6,8 +6,75 @@
...
+        """
+
+        request.user = None
+
+        # get auth_header from request
+        auth_header = authentication.get_authorization_header(request).split()
+        auth_header_prefix = self.auth_header_prefix.lower()
+
+        # Do not attempt to authenticate
+        if not auth_header:
```
Comment (oagutu) [inline authors/apps/authentication/backends.py]: I think lines 26-43 can be put under a single if statement

**Review by oagutu — COMMENTED** [review]
@ValerieRono good job. Please look into the feedback given

File: authors/apps/authentication/tests/test_jwt.py (hunk 1 above)
Comment (ValerieRono) [reply authors/apps/authentication/tests/test_jwt.py]: the 2 are not the same, user_data contains the payload required when registering, i.e. username, email and password, whereas, login_data contains the payload requires when logging in i.e email and password

**Pull request merged**